sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataAggregator.java

```diff
@@ -0,0 +1,279 @@
+                    updateOrPublishBatch(null, true);
+                });
+        }
+
+        /**
+         * Request a number of {@link EventDataBatch}.
+         *
+         * @param n Number of batches requested.
+         */
+        @Override
+        public void request(long n) {
+            if (!Operators.validate(n)) {
+                return;
+            }
+
+            Operators.addCap(REQUESTED, this, n);
+            subscription.request(n);
+        }
+
+        /**
+         * Cancels the subscription upstream.
+         */
+        @Override
+        public void cancel() {
+            // Do not keep requesting more events upstream
+            subscription.cancel();
+
+            updateOrPublishBatch(null, true);
+            downstream.onComplete();
+            disposable.dispose();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (subscription != null) {
+                logger.warning("Subscription was already set. Cancelling existing subscription.");
+                subscription.cancel();
+            } else {
+                subscription = s;
+                downstream.onSubscribe(this);
+            }
+        }
+
+        @Override
+        public void onNext(EventData eventData) {
+            updateOrPublishBatch(eventData, false);
+            eventSink.emitNext(1L, Sinks.EmitFailureHandler.FAIL_FAST);
+
+            final long left = REQUESTED.get(this);
+            if (left > 0) {
+                subscription.request(1L);
+            }
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            if (!isCompleted.compareAndSet(false, true)) {
+                Operators.onErrorDropped(t, downstream.currentContext());
+                return;
+            }
+
+            updateOrPublishBatch(null, true);
+            downstream.onError(t);
+        }
+
+        /**
+         * Upstream signals a completion.
+         */
+        @Override
+        public void onComplete() {
+            if (isCompleted.compareAndSet(false, true)) {
+                updateOrPublishBatch(null, true);
+                downstream.onComplete();
+            }
+        }
+
+        /**
+         * @param eventData EventData to add to or null if there are no events to add to the batch.
+         * @param alwaysPublish {@code true} to always push batch downstream. {@code false}, otherwise.
+         */
+        private void updateOrPublishBatch(EventData eventData, boolean alwaysPublish) {
+            if (alwaysPublish) {
+                publishDownstream();
+                return;
+            } else if (eventData == null) {
+                return;
+            }
+
+            boolean added;
+            synchronized (lock) {
+                added = currentBatch.tryAdd(eventData);
+            }
+
+            if (added) {
+                return;
+            }
+
+            publishDownstream();
+
+            synchronized (lock) {
```

Comment:
I understand why we have two tryAdd() here. But if we separate the lock block for two tryAdd(), between `publishDownstream()` and the second `tryAdd()`, I think it is possible that other threads add eventData into batch which cause the second `tryAdd()` fails. Maybe we need lock from the first `tryAdd()` until the second `tryAdd()`?

Reply:
Updated. :)